server/map_rpc_server.go

```diff
@@ -354,66 +354,98 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 
 	var newRoot *trillian.SignedMapRoot
 	err = t.registry.MapStorage.ReadWriteTransaction(ctx, tree, func(ctx context.Context, tx storage.MapTreeTX) error {
-		writeRev, err := tx.WriteRevision(ctx)
+		writeRev, err := t.writeRevision(ctx, tree, tx, req.Revision)
 		if err != nil {
 			return err
 		}
-		if rev := req.Revision; rev != 0 && writeRev != rev {
-			return status.Errorf(codes.FailedPrecondition, "can't write to revision %v", rev)
+
+		hkv, err := t.writeLeaves(ctx, tree, hasher, tx, req.Leaves, writeRev)
+		if err != nil {
+			return err
 		}
-		glog.V(2).Infof("%v: Writing at revision %v", mapID, writeRev)
-		txRunner := t.newTXRunner(tree, tx)
-		smtWriter, err := merkle.NewSparseMerkleTreeWriter(ctx, req.MapId, writeRev, hasher, txRunner)
+
+		newRoot, err = t.updateTree(ctx, tree, hasher, tx, hkv, req.Metadata, writeRev)
 		if err != nil {
 			return err
 		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &trillian.SetMapLeavesResponse{MapRoot: newRoot}, nil
+}
 
-		hkv := make([]merkle.HashKeyValue, 0, len(req.Leaves))
-		for _, l := range req.Leaves {
-			if err := checkIndexSize(l.Index, hasher); err != nil {
-				return err
-			}
-			l.LeafHash = hasher.HashLeaf(mapID, l.Index, l.LeafValue)
+// writeRevision ensures the desired write revision is available and claim a lock on this revision
```

Comment:
The lock claiming behavior is DB-specific and doesn't necessarily take place. In some there is optimistic / multi-versioned concurrency control, where the fact of revision not progressing is checked when the TX is committed.

I suggest focusing on the logical meaning of this "locking", e.g.:

```golang
// writeRevision returns the map revision that the passed-in transaction writes
// to. It verifies that this revision corresponds to the expected value, which
// helps sorting out conflicting transactions, should they occur.
```

Reply:
I feel like it should be a requirement of our data layer that it won't allow multiple updates to be stored under the same revision number. Which data layers allow transactions with conflicting revision IDs to be applied?

Reply:
Sure, I'm just saying that the comment probably shouldn't have details on how the conflict resolution works, because it depends on the database used within the storage layer. MySQL has locks, but something like CockroachDB is optimistic (I think).

Reply:
From CockroachDB [docs](https://www.cockroachlabs.com/blog/serializable-lockless-distributed-isolation-cockroachdb/):

> **Lockless**: Operations execute without taking locks on resources. Correctness is enforced by aborting transactions which would violate either serializability or strict scheduling.

Reply:
> I feel like it should be a requirement of our data layer that it won't allow multiple updates to be stored under the same revision number.

I think it is true now, multiple updates aren't possible.

Unrelated to this PR: In a massively parallel world though, you can imagine multiple parallel shards (split by leaf hash) writing leaves to the same revision, which is then "finalized" by a "master".

Reply:
Leaving as-is for now. The docs are saying what & why the code is at this revision. Currently this is correct and is what prevents us from creating conflicting revision states in the DB. As we add lock-free or massively parallel clever stuff, this will need to be revisited.